File: web/tests/device-session-retry.test.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { DeviceSessionRetry } from '../src/lib/api/DeviceSessionRetry.ts';

test('expired device sessions renew once across concurrent and late responses', async () => {
  let token = 'expired';
  let renewals = 0;
  const retry = new DeviceSessionRetry(() => token);
  retry.renew = async () => {
    renewals++;
    await new Promise(resolve => setTimeout(resolve, 5));
    token = 'fresh';
  };
  const run = delay => retry.run(async used => {
    if (used === 'expired') await new Promise(resolve => setTimeout(resolve, delay));
    return { status: used === 'fresh' ? 204 : 401 };
  });
  const results = await Promise.all([...Array.from({ length: 12 }, () => run(0)), run(25)]);
  assert.ok(results.every(r => r.status === 204));
  assert.equal(renewals, 1);
});

test('forbidden, throttled and delivered responses are never replayed', async () => {
  const retry = new DeviceSessionRetry(() => 'valid');
  retry.renew = async () => assert.fail('must not renew');
  for (const status of [204, 403, 429, 500]) {
    let calls = 0;
    assert.equal((await retry.run(async () => { calls++; return { status }; })).status, status);
    assert.equal(calls, 1);
  }
});

test('renewal failure can recover later, while a rejected renewed token stops after one retry', async () => {
  const retry = new DeviceSessionRetry(() => 'token');
  retry.renew = async () => { throw new Error('offline'); };
  await assert.rejects(retry.run(async () => ({ status: 401 })), /offline/);
  let calls = 0;
  retry.renew = async () => {};
  const result = await retry.run(async () => { calls++; return { status: 401 }; });
  assert.equal(result.status, 401);
  assert.equal(calls, 2);
});

File: app/test/device_session_retry_test.dart
import 'package:app/api/client.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;

void main() {
  tearDown(() {
    setDeviceAccessToken(null);
    setDeviceSessionRenewal(null);
  });

  test('concurrent expiry and a late 401 share one renewal', () async {
    setDeviceAccessToken('expired');
    var renewals = 0;
    setDeviceSessionRenewal(() async {
      renewals++;
      await Future<void>.delayed(const Duration(milliseconds: 5));
      setDeviceAccessToken('fresh');
    });
    Future<http.Response> run(int delay) => withDeviceAuthRetry(() async {
      final used = deviceApiHeaders['Authorization'];
      if (used == 'Bearer expired') {
        await Future<void>.delayed(Duration(milliseconds: delay));
      }
      return http.Response('', used == 'Bearer fresh' ? 204 : 401);
    });
    final responses = await Future.wait([
      ...List.generate(12, (_) => run(0)),
      run(25),
    ]);
    expect(responses.every((r) => r.statusCode == 204), isTrue);
    expect(renewals, 1);
  });

  test('business errors and delivered messages are never replayed', () async {
    setDeviceSessionRenewal(() async => fail('must not renew'));
    for (final status in [204, 403, 429, 500]) {
      var calls = 0;
      final result = await withDeviceAuthRetry(() async {
        calls++;
        return http.Response('', status);
      });
      expect(result.statusCode, status);
      expect(calls, 1);
    }
  });

  test('network recovery remains possible and retries are bounded', () async {
    setDeviceAccessToken('expired');
    setDeviceSessionRenewal(() async => throw Exception('offline'));
    await expectLater(
      withDeviceAuthRetry(() async => http.Response('', 401)),
      throwsA(isA<Exception>()),
    );
    var calls = 0;
    setDeviceSessionRenewal(() async {});
    final result = await withDeviceAuthRetry(() async {
      calls++;
      return http.Response('', 401);
    });
    expect(result.statusCode, 401);
    expect(calls, 2);
  });
}
